accord-core/src/test/java/accord/impl/basic/Journal.java

```diff
@@ -0,0 +1,361 @@
+    private final Map<TxnId, Map<MessageType, Message>> writes = new HashMap<>();
+    private Node node;
+    
+    public void start(Node node)
+    {
+        this.node = node;
+        node.scheduler().recurring(this, 1, TimeUnit.MILLISECONDS);
+    }
+
+    public void shutdown()
+    {
+        this.node = null;
+    }
+    
+    @Override
+    public void handle(LocalRequest<?> message, Node node)
+    {
+        if (message.type().hasSideEffects())
+        {
+            // enqueue
+            unframedRequests.add(new RequestContext(message, () -> node.scheduler().now(() -> message.process(node))));
+            return;
+        }
+        message.process(node);
+    }
+
+    public void handle(Request request, Node.Id from, ReplyContext replyContext)
+    {
+        if (request.type() != null && request.type().hasSideEffects())
+        {
+            // enqueue
+            unframedRequests.add(new RequestContext(request, () -> node.receive(request, from, replyContext)));
+            return;
+        }
+        node.receive(request, from, replyContext);
+    }
+
+    private void save(Message request)
+    {
+        MessageType type = request.type();
+        TxnIdProvider provider = typeToProvider.get(type);
+        Invariants.nonNull(provider, "Unknown type %s: %s", type, request);
+        TxnId txnId = provider.txnId(request);
+        writes.computeIfAbsent(txnId, ignore -> new HashMap<>()).put(type, request);
+    }
+
+    public SerializerSupport.MessageProvider makeMessageProvider(TxnId txnId)
+    {
+        return new MessageProvider(txnId, writes.getOrDefault(txnId, Map.of()));
+    }
+
+    @Override
+    public void run()
+    {
+        if (this.node == null)
+            return;
+        try
+        {
+            doRun();
+        }
+        catch (Throwable t)
+        {
+            node.agent().onUncaughtException(t);
+        }
+    }
+
+    private void doRun()
+    {
+        ArrayList<RequestContext> requests = null;
+        // check to see if any pending epochs are in
+        waitForEpochs.sort(null);
+        for (int i = 0; i < waitForEpochs.size(); i++)
+        {
+            long waitForEpoch = waitForEpochs.getLong(i);
+            if (!node.topology().hasEpoch(waitForEpoch))
+                break;
+            List<RequestContext> delayed = delayedRequests.remove(waitForEpoch);
+            if (null == requests) requests = new ArrayList<>(delayed.size());
+            requests.addAll(delayed);
+        }
+        waitForEpochs.removeIfLong(epoch -> !delayedRequests.containsKey(epoch));
+        
+        // for anything queued, put into the pending epochs or schedule
+        RequestContext request;
+        while (null != (request = unframedRequests.poll()))
+        {
+            long waitForEpoch = request.waitForEpoch;
+            if (waitForEpoch != 0 && !node.topology().hasEpoch(waitForEpoch))
```

Comment:
`waitForEpoch != 0 && ` was also fixed in the `AccordJournal` though not sure we ever hit that... we did in burn test and know `epoch=0` isn't valid (its the empty state)